modules/platforms/dotnet/Apache.Ignite.Core/Impl/Unmanaged/UnmanagedThread.cs

```diff
@@ -0,0 +1,214 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apache.Ignite.Core.Impl.Unmanaged
+{
+    using System;
+    using System.Diagnostics;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Runtime.InteropServices;
+
+    /// <summary>
+    /// Unmanaged thread utils.
+    /// </summary>
+    internal static class UnmanagedThread
+    {
+        /// <summary>
+        /// Delegate for <see cref="SetThreadExitCallback"/>.
+        /// </summary>
+        /// <param name="threadLocalValue">Value from <see cref="EnableCurrentThreadExitEvent"/></param>
+        public delegate void ThreadExitCallback(IntPtr threadLocalValue);
+
+        /// <summary>
+        /// Sets the thread exit callback, and returns an id to pass to <see cref="EnableCurrentThreadExitEvent"/>.
+        /// </summary>
+        /// <param name="callbackPtr">
+        /// Pointer to a callback function that matches <see cref="ThreadExitCallback"/>.
+        /// </param>
+        public static unsafe int SetThreadExitCallback(IntPtr callbackPtr)
+        {
+            Debug.Assert(callbackPtr != IntPtr.Zero);
+
+            if (Os.IsWindows)
+            {
+                var res = NativeMethodsWindows.FlsAlloc(callbackPtr);
+
+                if (res == NativeMethodsWindows.FLS_OUT_OF_INDEXES)
+                {
+                    throw new InvalidOperationException("FlsAlloc failed: " + Marshal.GetLastWin32Error());
+                }
+
+                return res;
+            }
+
+            if (Os.IsMacOs)
+            {
+                int tlsIndex;
+                var res = NativeMethodsMacOs.pthread_key_create(new IntPtr(&tlsIndex), callbackPtr);
+
+                NativeMethodsLinux.CheckResult(res);
+
+                return tlsIndex;
+            }
+
+            if (Os.IsLinux)
+            {
+                int tlsIndex;
+                var res = NativeMethodsLinux.pthread_key_create(new IntPtr(&tlsIndex), callbackPtr);
+
+                NativeMethodsLinux.CheckResult(res);
+
+                return tlsIndex;
+            }
+
+            throw new InvalidOperationException("Unsupported OS: " + Environment.OSVersion);
+        }
+
+        /// <summary>
+        /// Removes thread exit callback that has been set with <see cref="SetThreadExitCallback"/>.
+        /// </summary>
+        /// <param name="callbackId">Callback id returned from <see cref="SetThreadExitCallback"/>.</param>
+        public static void RemoveThreadExitCallback(int callbackId)
```

Comment:
Actually, it not only removes callback, but also calls it in case of Windows: https://docs.microsoft.com/en-us/windows/win32/api/winnt/nc-winnt-pfls_callback_function . Need a test for it.

Reply:
Good point. However:

1) This method is only called from tests - we don't need it during actual Ignite usage since we don't support JVM destruction. 

2) It is fine to call `DetachCurrentThread` multiple times, so it is fine anyway.